plugins/lookup/etcd3.py

```diff
@@ -0,0 +1,231 @@
+        msg: "{{ lookup('community.general.etcd3', 'foo', 'bar', 'baz') }}"
+
+    - name: "look for a key prefix"
+      debug:
+        msg: "{{ lookup('community.general.etcd3', '/foo/bar', prefix=True) }}"
+
+    - name: "connect to etcd3 with a client certificate"
+      debug:
+        msg: "{{ lookup('community.general.etcd3', 'foo/bar', cert_cert='/etc/ssl/etcd/client.pem', cert_key='/etc/ssl/etcd/client.key') }}"
+'''
+
+RETURN = '''
+    _raw:
+        description:
+        - List of keys and associated values.
+        type: list
+        elements: dict
+        contains:
+            key:
+                description: The element's key.
+                type: str
+            value:
+                description: The element's value.
+                type: str
+'''
+
+import re
+
+from ansible.plugins.lookup import LookupBase
+from ansible.utils.display import Display
+from ansible.module_utils.basic import missing_required_lib
+from ansible.module_utils._text import to_native
+from ansible.plugins.lookup import LookupBase
+from ansible.errors import AnsibleError, AnsibleLookupError
+
+try:
+    import etcd3
+    HAS_ETCD = True
+except ImportError:
+    HAS_ETCD = False
+
+display = Display()
+
+etcd3_cnx_opts = (
+    'host',
+    'port',
+    'ca_cert',
+    'cert_key',
+    'cert_cert',
+    'timeout',
+    'user',
+    'password',
+    # 'grpc_options' Etcd3Client() option currently not supported by lookup module (maybe in future ?)
+)
+
+
+def etcd3_client(client_params):
+    try:
+        etcd = etcd3.client(**client_params)
+        etcd.status()
+    except Exception as exp:
+        raise AnsibleLookupError('Cannot connect to etcd cluster: %s' % (to_native(exp)))
+    return etcd
+
+
+class LookupModule(LookupBase):
+
+    def run(self, terms, variables, **kwargs):
+
+        self.set_options(var_options=variables, direct=kwargs)
+
+        if not HAS_ETCD:
+            display.error(missing_required_lib('etcd3'))
+            return None
+
+        # create the etcd3 connection parameters dict to pass to etcd3 class
+        client_params = {}
+
+        # etcd3 class expects host and port as connection parameters, so endpoints
+        # must be mangled a bit to fit in this scheme.
+        # so here we use a regex to extract server and port
+        match = re.compile(
+            r'^(https?://)?(?P<host>(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})|([-_\d\w\.]+))(:(?P<port>\d{1,5}))?/?$'
+        ).match(self.get_option('endpoints'))
+        if match:
+            if match.group('host'):
+                client_params['host'] = match.group('host')
+            if match.group('port'):
+                client_params['port'] = match.group('port')
+
+        for opt in etcd3_cnx_opts:
+            if self.get_option(opt):
+                client_params[opt] = self.get_option(opt)
+
+        cnx_log = dict(client_params)
+        if 'password' in cnx_log.keys():
+            cnx_log['password'] = '<password obfuscated by Ansible>'
```

Comment:
How about:
```suggestion
            cnx_log['password'] = '<redacted>'
```
That makes the log message shorter (and thus easier to read).